### PR TITLE
implement ConnPrepareContext/StmtQueryContext/StmtExecContext interfaces

### LINKED ISSUE #1046: stmt.QueryRowContext doesn't respect canceled context
Working on a project using https://github.com/jmoiron/sqlx, I ran into this issue when using `PrepareContext()`/`QueryRowContext` with a context with timeout. 

The following test case exercises the problem
```
func TestIssue1046(t *testing.T) {
	ctxTimeout := time.Second * 2
	sleepSQL := `SELECT pg_sleep(10) AS id`

	db := openTestConn(t)
	defer db.Close()

	ctx, cancel := context.WithTimeout(context.Background(), ctxTimeout)
	defer cancel()

	stmt, err := db.PrepareContext(ctx, sleepSQL)
	if err != nil {
		t.Fatal(err)
	}

	var d []uint8
	err = stmt.QueryRowContext(ctx).Scan(&d)
	dl, _ := ctx.Deadline()
	since := time.Since(dl)
	if since > ctxTimeout {
		t.Logf("FAIL %s: query returned after context deadline: %v\n", t.Name(), since)
		t.Fail()
	}
	expectedErr := &Error{Message: "canceling statement due to user request"}
	if err == nil || err.Error() != expectedErr.Error() {
		t.Logf("ctx.Err(): [%T]%+v\n", ctx.Err(), ctx.Err())
		t.Logf("got err: [%T] %+v expected err: [%T] %+v", err, err, expectedErr, expectedErr)
		t.Fail()
	}
}
```

### PR DESCRIPTION
**Pull request by michaelshobbs**
See #1046 for more context. (pun intended)

This is almost a line for line copy of #921 with test cases (thanks @kylejbrock). Please let me know what else needs to be done to get this merged. 

Also dropped ci testing of unsupported versions (9.5 and 9.4) per pg docs [here](https://www.postgresql.org/docs/) and added ci testing of versions 11, 12, and 13.

closes #921 
closes #1046

### CODE REVIEW

**Comment by michaelshobbs** [comment]
@otan @mjibson please let me know if i can be of any more assistance in getting this merged

**Comment by michaelshobbs** [comment]
@otan @mjibson
i know `pgx` is the preferred path forward. however, i also know there are many folks still using `pq` and this addition would certainly be a value add for those users. please let me know how i can help move this forward. 

thank you 🙏

**Review by otan — COMMENTED** [review]
My gut with this is the same as https://github.com/lib/pq/pull/921#discussion_r356281669 -- there are too many goroutines and channels for me to feel comfortable. But given that we already do this for the `conn` implementation, maybe we're ok.

Some testing changes needed.

File: conn_test.go [hunk 1]
```diff
@@ -1806,3 +1806,160 @@ func TestCopyInStmtAffectedRows(t *testing.T) {
...
+			if cancel != nil {
+				defer cancel()
+			}
+			stmt, err := db.PrepareContext(ctx, tt.sql)
+			if err != nil {
+				t.Fatal(err)
+			}
+			_, err = stmt.ExecContext(ctx)
+			if ((err != nil) != (tt.err != nil)) && (err.Error() != tt.err.Error()) {
+				fmt.Printf("err: %+v\n", err)
```
[the remark this answers is not in the document]
Comment (michaelshobbs) [reply conn_test.go]: fixed here: c7f67ff

File: conn_test.go [hunk 2]
```diff
@@ -1806,3 +1806,160 @@ func TestCopyInStmtAffectedRows(t *testing.T) {
...
+			ctx, cancel := tt.ctx()
+			if cancel != nil {
+				defer cancel()
+			}
+			stmt, err := db.PrepareContext(ctx, tt.sql)
+			if err != nil {
+				t.Fatal(err)
+			}
+			_, err = stmt.ExecContext(ctx)
+			if ((err != nil) != (tt.err != nil)) && (err.Error() != tt.err.Error()) {
```
[the remark this answers is not in the document]
Comment (michaelshobbs) [reply conn_test.go]: my intent was to catch an errantly configured test where we got an error but didn't expect it and/or didn't get one but did expect it. i think this (55727bb) clarifies that but let me know if not.

File: conn_test.go [hunk 3]
```diff
@@ -1806,3 +1806,160 @@ func TestCopyInStmtAffectedRows(t *testing.T) {
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := tt.ctx()
+			if cancel != nil {
+				defer cancel()
+			}
+			_, err := db.PrepareContext(ctx, tt.sql)
+			if ((err != nil) != (tt.err != nil)) && (err.Error() != tt.err.Error()) {
```
[the remark this answers is not in the document]
Comment (michaelshobbs) [reply conn_test.go]: 55727bb

File: conn_go18.go [hunk 4]
```diff
@@ -172,3 +180,67 @@ func (cn *conn) cancel(ctx context.Context) error {
...
+	if done := ctx.Done(); done != nil {
+		finished := make(chan struct{})
+		go func() {
+			select {
+			case <-done:
+				// At this point the function level context is canceled,
+				// so it must not be used for the additional network
+				// request to cancel the query.
+				// Create a new context to pass into the dial.
+				ctxCancel, cancel := context.WithTimeout(context.Background(), time.Second*10)
```
[the remark this answers is not in the document]
Comment (michaelshobbs) [reply conn_go18.go:226]: > a 10 second timeout to cancel the query feels very arbitrary.

i agree. however, this is precisely what we do in `conn.watchCancel()` on L120

https://github.com/lib/pq/blob/8cb9b672461aac467ea514110abbdf9e65694c90/conn_go18.go#L120

looking into why i didn't use `st.cn.watchCancel()`now...

[the remark this answers is not in the document]
Comment (michaelshobbs) [reply conn_go18.go:226]: > should the statement's conn be reusable if we cancel on a stmt?

ah exactly this. my take was that a statement context timeout should not affect the underlying connection.

[the remark this answers is not in the document]
Comment (otan) [reply conn_go18.go:226]: can you add a comment to that effect? also, can you move the 10*time.Second to a constant?

[the remark this answers is not in the document]
Comment (michaelshobbs) [reply conn_go18.go]: done here: 94c98c3

**Comment by otan** [comment]
well looks like CI no longer works :\

**Comment by michaelshobbs** [comment]
Yea looks like they deprecated travis-ci.org in favor of travis-ci-com. not sure what needs to happen to move this project over. Let me know if I can help.

**Comment by michaelshobbs** [comment]
Looks like all you have to do is sign up on the .com?

https://blog.travis-ci.com/2021-05-07-orgshutdown

**Comment by otan** [comment]
i've tried that and resent the webhooks but no bueno, perhaps try `git commit --amend --reset-author -a` and `-git push --force` your commit to try retrigger it.

**Comment by michaelshobbs** [comment]
I was able to add my fork here: https://travis-ci.com/github/michaelshobbs/pq/builds/235721359

I assume that's because I own it. Who is still an admin on this repo? 😬

**Comment by michaelshobbs** [comment]
@otan @mjibson i'll give this a look this week

**Comment by michaelshobbs** [comment]
@otan @mjibson can one of you approve me for running workflows?

**Comment by otan** [comment]
gotcha

**Comment by DoubleDi** [comment]
Hi! 
Thank you very much for this PR. Just found out our code has the same problems with cancelling requests from prepared statements. Would be awesome to get this into master ASAP.

**Comment by DoubleDi** [comment]
Quick update @michaelshobbs. We have tested your patch (just replaced the go.mod to your fork) and it seems still not to work on prepared statements.

**Comment by michaelshobbs** [comment]
@DoubleDi can you provide some code that demonstrates the incorrect behavior? this PR includes `canceling statement due to user request` test assertions on `stmt.QueryContext()`, and `stmt.ExecContext()`. i'd like to better understand where we might be missing something here. thanks 🙏

**Comment by michaelshobbs** [comment]
@otan @mjibson https://github.com/lib/pq/pull/1054

**Comment by michaelshobbs** [comment]
@otan @mjibson rebased master. this should be good to go now

**Comment by DoubleDi** [comment]
@otan thanks! Please also tag a new release

**Comment by otan** [comment]
v1.10.3, you have to be a *tad* more patient than 90s :P

**Comment by DoubleDi** [comment]
thank you so much!

**Pull request merged**